lib/rucio/core/request.py

```diff
@@ -1534,6 +1534,14 @@ def add_monitor_message(new_state, request, additional_fields, session=None):
         transfer_status = 'transfer-%s' % new_state.name
     transfer_status = transfer_status.lower()
 
+    stmt = select(
+        models.DataIdentifier.datatype
+    ).where(
+        models.DataIdentifier.scope == request['scope'],
+        models.DataIdentifier.name == request['name'],
+    )
+    datatype = session.execute(stmt).scalar_one_or_none()
+
```

Comment:
I'm not sure this is a good idea. I cannot put `get_datatype_from_did_table` into a common place (dids.py) for reasons I explain in the commit message. The function is not complicated at all (don't know why you say it's a monster). I don't think it's justified in this particular case to move the query to a separate function which nobody else will use except this code.

Reply:
While this function "only" creates a message for hermes, it does this with a lot of logic
- It creates the `transfer-status` based on the request type and the new_state name
  - It makes the `transfer-status` lower case
- It selects the data-type (or None) from the DIDs table, based on the defined primary-key
- It constructs the message with 31 fields, some of them are None for any reason apperently
- It then updates the default message values with some additional fields, or adds them if they dont exist
- It then sets some of the values, which are by default None, to a value based on logic which is implemented with 7 if statements and one for loop
- And it adds the message in the end

I think we mix several use-cases together:
- Creating and adding a message:
  - These are two different things, so why don't we have a `create_hermes_message` function?
- Creating the `transfer_status`:
  - This has nothing to do with the message creation, just with the adding a message. So why is there no `get_hermes_transfer_status`?

From my POV, the `add_monitor_message` should look like:
```python
def get_hermes_transfer_status(new_state: str, request_type: ... = None) -> str:
    if request_type:
        transfer_status = '%s-%s' % (request_type.name, new_state)
    else:
        transfer_status = 'transfer-%s' % new_state
    return transfer_status.lower()


def create_hermes_message(...):
    pass


@read_session
def add_monitor_message(new_state, request, additional_fields, session=None):
    transfer_status = get_hermes_transfer_status(new_state, request['request_type'])
    message = create_hermes_message(request, additional_fields, session=session)  # No new_state

    add_message(transfer_status, message, session=session)
```

And now we can test the creation of the transfer_status, the creation of the message and the add message independently. If there is an error in the message, I only look into `create_hermes_message`, not in `get_hermes_transfer_status`. In the current _petit_ monster, this would've cluttered my mind too.

To the `get_datatype_from_did_table`: You can add the function in this file and another one in the other file. It is not about code duplication (in this case), but readability. If I have to change a field in the message not related to the `datatype`, I don't want to read the code to get the `datatype`. A function with a good name therefore allows me to skip this part, and look into it if it's important.

Reply:
The "improvement" you propose is very subjective and I don't agree that it makes things better. I consider it complicates the flow of understanding the function. This is a very simple function without loops and without complex state changes which does only one job: correctly formatting a "message" dictionary.

Reply:
Ok, let us keep it as it is.